velox/vector/arrow/Bridge.cpp

```diff
@@ -686,27 +861,92 @@ RowVectorPtr createRowVector(
   childrenVector.reserve(arrowArray.n_children);
 
   for (size_t i = 0; i < arrowArray.n_children; ++i) {
-    childrenVector.emplace_back(
-        isViewer
-            ? importFromArrowAsViewer(
-                  *arrowSchema.children[i], *arrowArray.children[i], pool)
-            : importFromArrowAsOwner(
-                  *arrowSchema.children[i], *arrowArray.children[i], pool));
+    childrenVector.push_back(importFromArrowImpl(
+        *arrowSchema.children[i], *arrowArray.children[i], pool, isViewer));
   }
   return std::make_shared<RowVector>(
       pool,
       rowType,
       nulls,
       arrowArray.length,
       std::move(childrenVector),
-      arrowArray.null_count == -1
-          ? std::nullopt
-          : std::optional<int64_t>(arrowArray.null_count));
+      optionalNullCount(arrowArray.null_count));
 }
 
-VectorPtr importFromArrowImpl(
+BufferPtr computeSizes(
+    const vector_size_t* offsets,
+    int64_t length,
+    memory::MemoryPool* pool) {
+  auto sizesBuf = AlignedBuffer::allocate<vector_size_t>(length, pool);
+  auto sizes = sizesBuf->asMutable<vector_size_t>();
+  for (int64_t i = 0; i < length; ++i) {
+    sizes[i] = offsets[i + 1] - offsets[i];
+  }
+  return sizesBuf;
+}
+
+ArrayVectorPtr createArrayVector(
+    memory::MemoryPool* pool,
+    const TypePtr& type,
+    BufferPtr nulls,
     const ArrowSchema& arrowSchema,
     const ArrowArray& arrowArray,
+    bool isViewer,
+    WrapInBufferViewFunc wrapInBufferView) {
+  static_assert(sizeof(vector_size_t) == sizeof(int32_t));
+  VELOX_CHECK_EQ(arrowArray.n_buffers, 2);
+  VELOX_CHECK_EQ(arrowArray.n_children, 1);
+  auto offsets = wrapInBufferView(
+      arrowArray.buffers[1], (arrowArray.length + 1) * sizeof(vector_size_t));
+  auto sizes =
+      computeSizes(offsets->as<vector_size_t>(), arrowArray.length, pool);
+  auto elements = importFromArrowImpl(
+      *arrowSchema.children[0], *arrowArray.children[0], pool, isViewer);
+  return std::make_shared<ArrayVector>(
+      pool,
+      type,
+      std::move(nulls),
+      arrowArray.length,
+      std::move(offsets),
+      std::move(sizes),
+      std::move(elements),
+      optionalNullCount(arrowArray.null_count));
+}
+
+MapVectorPtr createMapVector(
+    memory::MemoryPool* pool,
+    const TypePtr& type,
+    BufferPtr nulls,
+    const ArrowSchema& arrowSchema,
+    const ArrowArray& arrowArray,
+    bool isViewer,
+    WrapInBufferViewFunc wrapInBufferView) {
+  VELOX_CHECK_EQ(arrowArray.n_buffers, 2);
+  VELOX_CHECK_EQ(arrowArray.n_children, 1);
+  auto offsets = wrapInBufferView(
+      arrowArray.buffers[1], (arrowArray.length + 1) * sizeof(vector_size_t));
+  auto sizes =
+      computeSizes(offsets->as<vector_size_t>(), arrowArray.length, pool);
+  auto entries = importFromArrowImpl(
+      *arrowSchema.children[0], *arrowArray.children[0], pool, isViewer);
+  VELOX_CHECK(entries->type()->isRow());
+  const auto& rows = *entries->asUnchecked<RowVector>();
```

Comment:
could use a tiny comment here explaining that Arrow wraps keys and values into a row/struct